### PR TITLE
audioreactive update (align with MoonMod code)

### PR DESCRIPTION
**Pull request by softhack007**
This update provides bug fixing & minor improvements for audioreactive usermod. It is based on the "MoonMod" version, but leaves out a few experimental features, like mic profiles and experimental "I2S slave mode" support.

@blazoncek would be nice if you can do a quick code review before we merge this PR.

**Changes to audio drivers:**
- small improvements for analog input driver
- Better handling of PDM and Line-In
- Bugfixes for ES7243 (allocateMultiplePins() needed for I2C part)
- More error messages for ES7243
- sample scaling (needed for sources that provide full scale (16bit) samples)

**Changes in audiorective usermod core:**
* align SR_DEBUG with WLED_DEBUG
* optional bandpass filter (needed for PDM mics)
* sample scaling for PDM and Line-In
* small improvements for analog input
* bugfixes and small performance improvements
* code for FFT task refactored, for better readablity. Introduces separate functions for filtering and post-processing
* small improvement for beat detection
* default mic settings can be configured at compile time
* correct mic type if MCU does not support PDM or ADC
* hide analog PIN config if not supported by MCU

### CODE REVIEW

File: usermods/audioreactive/audio_reactive.h [hunk 1]
```diff
@@ -106,68 +106,86 @@ static void autoResetPeak(void);     // peak auto-reset function
...
-
-// These are the input and output vectors.  Input vectors receive computed results from FFT.
-static float vReal[samplesFFT] = {0.0f};       // FFT sample inputs / freq output -  these are our raw result bins
-static float vImag[samplesFFT] = {0.0f};       // imaginary parts
-
 // the following are observed values, supported by a bit of "educated guessing"
 //#define FFT_DOWNSCALE 0.65f                             // 20kHz - downscaling factor for FFT results - "Flat-Top" window @20Khz, old freq channels 
 #define FFT_DOWNSCALE 0.46f                             // downscaling factor for FFT results - for "Flat-Top" window @22Khz, new freq channels
-#define LOG_256  5.54517744
+#define LOG_256  5.54517744                             // log(256)
```
Comment (blazoncek) [inline usermods/audioreactive/audio_reactive.h:155]: Add `f` for float.

File: usermods/audioreactive/audio_reactive.h [hunk 2]
```diff
@@ -329,9 +363,67 @@ void FFTcode(void * parameter)
...
+  //constexpr float alpha = 0.03f;   // 110Hz
+  constexpr float alpha = 0.0225f; // 80hz
+  //constexpr float alpha = 0.01693f;// 60hz
+  // high frequency cutoff  parameter
+  //constexpr float beta1 = 0.75;    // 11Khz
+  //constexpr float beta1 = 0.82;    // 15Khz
+  //constexpr float beta1 = 0.8285;  // 18Khz
+  constexpr float beta1 = 0.85;   // 20Khz
+
+  constexpr float beta2 = (1.0f - beta1) / 2.0;
```
Comment (blazoncek) [inline usermods/audioreactive/audio_reactive.h:404]: Add `f`

File: usermods/audioreactive/audio_reactive.h [hunk 3]
```diff
@@ -538,15 +626,14 @@ class AudioReactive : public Usermod {
     // variables used by getSample() and agcAvg()
     int16_t  micIn = 0;           // Current sample starts with negative values and large values, which is why it's 16 bit signed
     double   sampleMax = 0.0;     // Max sample over a few seconds. Needed for AGC controler.
-    float    micLev = 0.0f;       // Used to convert returned value to have '0' as minimum. A leveller
+    double   micLev = 0.0f;       // Used to convert returned value to have '0' as minimum. A leveller
```
Comment (blazoncek) [inline usermods/audioreactive/audio_reactive.h:629]: If double, no `f` required.

File: usermods/audioreactive/audio_reactive.h [hunk 4]
```diff
@@ -579,7 +666,7 @@ class AudioReactive : public Usermod {
       Serial.print("micReal:");     Serial.print(micDataReal); Serial.print("\t");
       Serial.print("volumeSmth:");  Serial.print(volumeSmth);  Serial.print("\t");
       //Serial.print("volumeRaw:");   Serial.print(volumeRaw);   Serial.print("\t");
-      //Serial.print("DC_Level:");    Serial.print(micLev);      Serial.print("\t");
+      Serial.print("DC_Level:");    Serial.print(micLev);      Serial.print("\t");
```
Comment (blazoncek) [inline usermods/audioreactive/audio_reactive.h:669]: Do we really need Serial or would we be better of converting to DEBUG_PRIINT?

**Review by blazoncek — APPROVED** [review]
I would much rather see AR_DMTYPE instead of SR_DMTYPE (and others) but I understand the source. 😉 
Unfortunately your magic is beyond my skill level so I only touched the surface.

File: usermods/audioreactive/audio_reactive.h (hunk 4 above)
Comment (softhack007) [reply usermods/audioreactive/audio_reactive.h:669]: Good point. The intention  is that this is used with Arduino Serial Plotter, so works best if no other debug output comes on serial. But surely will also work over network with DEBUGOUT. Maybe best if I create a macro like PLOT_PRINT that is defined like DEBUG_PRINT.

File: usermods/audioreactive/audio_reactive.h (hunk 2 above)
Comment (softhack007) [reply usermods/audioreactive/audio_reactive.h:414]: done, thanks.

**Comment by softhack007** [comment]
Hi @blazoncek, please cross-check my changes to the readme.md. 
If OK for you, please hit "merge" ;-)

**Comment by softhack007** [comment]
I'm merging this now, before source codes starts to diverge..  @blazoncek please feel free to modify the usermod readme.md in case you want something changed/added.

**Pull request merged**